File: test/controllers/api/v2/system_errata_controller_test.rb
# encoding: utf-8
#
# Copyright 2014 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

require "katello_test_helper"

module Katello
  class Api::V2::SystemErrataControllerTest < ActionController::TestCase
    include Support::ForemanTasks::Task

    def self.before_suite
      models = ["System"]
      disable_glue_layers(["Candlepin", "Pulp", "ElasticSearch"], models)
      super
    end

    def permissions
      @view_permission = :view_content_hosts
      @create_permission = :create_content_hosts
      @update_permission = :edit_content_hosts
      @destroy_permission = :destroy_content_hosts
    end

    def setup
      setup_controller_defaults_api
      login_user(User.find(users(:admin)))
      @request.env['HTTP_ACCEPT'] = 'application/json'

      @system = katello_systems(:simple_server)
      permissions
    end

    def test_apply
      assert_async_task ::Actions::Katello::System::Erratum::Install do |system, errata|
        system.id == @system.id && errata == %w(RHSA-1999-1231)
      end

      put :apply, :system_id => @system.uuid, :errata_ids => %w(RHSA-1999-1231)

      assert_response :success
    end

    def test_apply_unknown_errata
      put :apply, :system_id => @system.uuid, :errata_ids => %w(non-existant-errata)
      assert_response 404
    end

    def test_apply_protected
      good_perms = [@update_permission]
      bad_perms = [@view_permission, @create_permission, @destroy_permission]

      assert_protected_action(:apply, good_perms, bad_perms) do
        put :apply, :system_id => @system.uuid, :errata_ids => %w(RHSA-1999-1231)
      end
    end
  end
end

File: engines/bastion/test/systems/content/system-errata.controller.test.js
/**
 * Copyright 2014 Red Hat, Inc.
 *
 * This software is licensed to you under the GNU General Public
 * License as published by the Free Software Foundation; either version
 * 2 of the License (GPLv2) or (at your option) any later version.
 * There is NO WARRANTY for this software, express or implied,
 * including the implied warranties of MERCHANTABILITY,
 * NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
 * have received a copy of GPLv2 along with this software; if not, see
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
 **/

describe('Controller: SystemErrataController', function() {
    var $scope, Nutupane, SystemErratum,
        mockSystem, mockTask, mockErratum;

    beforeEach(module('Bastion.systems', 'Bastion.test-mocks'));

    beforeEach(function() {
        mockErratum = {
            errata_id: "RHSA-1024"
        };
        mockSystem = {
            uuid: 5
        };
        mockTask = {
            pending: true,
            id: 7
        };
        Nutupane = function() {
            this.table = {
                showColumns: function() {},
                getSelected: function() {return [mockErratum];},
                selectAll: function() {}
            };
            this.get = function() {};
        };
        SystemErratum = {
            get: function() {return []},
            apply: function(errata, success) {
                success(mockTask);
                return mockTask
            }
        };
    });

    beforeEach(inject(function($controller, $rootScope) {
        $scope = $rootScope.$new();
        $scope.system = mockSystem;

        $controller('SystemErrataController', {$scope: $scope,
                                               SystemErratum: SystemErratum,
                                               Nutupane: Nutupane});
    }));

    it("Sets a table.", function() {
        expect($scope.errataTable).toBeTruthy();
    });

    it("provide a way to apply errata", function() {
        spyOn(SystemErratum, "apply").andCallThrough();
        spyOn($scope.errataTable, "selectAll");
        spyOn($scope, "transitionTo");
        $scope.applySelected();
        expect(SystemErratum.apply).toHaveBeenCalledWith({uuid: mockSystem.uuid, errata_ids: [mockErratum.errata_id]},
                                                         jasmine.any(Function));
        expect($scope.transitionTo).toHaveBeenCalledWith('systems.details.tasks.details', {taskId: mockTask.id});
        expect($scope.errataTable.selectAll).toHaveBeenCalledWith(false);
    });
});
